File: frontend/src/services/livekitService.js
/**
 * LiveKit Cloud Service
 * Handles all LiveKit Cloud integrations with optimized settings
 * for iPhone Safari and mobile-first approach
 */

import axios from 'axios';

const BACKEND_URL = process.env.REACT_APP_BACKEND_URL;
const LIVEKIT_URL = process.env.REACT_APP_LIVEKIT_URL;

class LiveKitService {
    constructor() {
        this.apiClient = axios.create({
            baseURL: BACKEND_URL,
            timeout: 10000,
            headers: {
                'Content-Type': 'application/json',
            },
        });
        
        // Add request interceptor for authentication
        this.apiClient.interceptors.request.use(
            (config) => {
                const token = localStorage.getItem('access_token');
                if (token) {
                    config.headers.Authorization = `Bearer ${token}`;
                }
                return config;
            },
            (error) => Promise.reject(error)
        );
    }

    /**
     * Generate LiveKit token for publisher (admin)
     */
    async generatePublisherToken(roomName, participantName = null, metadata = null) {
        try {
            const response = await this.apiClient.post('/api/livekit/token', {
                room_name: roomName,
                participant_type: 'publisher',
                participant_name: participantName,
                metadata: metadata || { role: 'admin', streaming: true }
            });

            return {
                token: response.data.token,
                roomName: response.data.room_name,
                participantIdentity: response.data.participant_identity,
                livekitUrl: response.data.livekit_url,
                expiresIn: response.data.expires_in
            };
        } catch (error) {
            console.error('Error generating publisher token:', error);
            throw new Error(`Failed to generate publisher token: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Generate LiveKit token for viewer (customer)
     */
    async generateViewerToken(roomName, participantName = null, metadata = null) {
        try {
            const response = await this.apiClient.post('/api/livekit/token', {
                room_name: roomName,
                participant_type: 'viewer',
                participant_name: participantName,
                metadata: metadata || { role: 'customer', viewing: true }
            });

            return {
                token: response.data.token,
                roomName: response.data.room_name,
                participantIdentity: response.data.participant_identity,
                livekitUrl: response.data.livekit_url,
                expiresIn: response.data.expires_in
            };
        } catch (error) {
            console.error('Error generating viewer token:', error);
            throw new Error(`Failed to generate viewer token: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Create a new LiveKit room
     */
    async createRoom(roomName, maxParticipants = 50, metadata = null) {
        try {
            const response = await this.apiClient.post('/api/livekit/room/create', {
                room_name: roomName,
                max_participants: maxParticipants,
                empty_timeout: 300, // 5 minutes
                metadata: metadata || { type: 'live_shopping', created_at: Date.now() }
            });

            return {
                roomName: response.data.room_name,
                sid: response.data.sid,
                maxParticipants: response.data.max_participants,
                numParticipants: response.data.num_participants,
                creationTime: response.data.creation_time,
                status: response.data.status
            };
        } catch (error) {
            console.error('Error creating room:', error);
            throw new Error(`Failed to create room: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Get list of active rooms
     */
    async getActiveRooms() {
        try {
            const response = await this.apiClient.get('/api/livekit/rooms');
            return response.data.rooms.map(room => ({
                roomName: room.room_name,
                sid: room.sid,
                numParticipants: room.num_participants,
                maxParticipants: room.max_participants,
                isLive: room.is_live,
                creationTime: room.creation_time,
                metadata: room.metadata
            }));
        } catch (error) {
            console.error('Error getting active rooms:', error);
            throw new Error(`Failed to get active rooms: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Get room information
     */
    async getRoomInfo(roomName) {
        try {
            const response = await this.apiClient.get(`/api/livekit/room/${roomName}`);
            return {
                room: response.data.room,
                participants: response.data.participants,
                isLive: response.data.is_live
            };
        } catch (error) {
            console.error('Error getting room info:', error);
            throw new Error(`Failed to get room info: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * End a room (admin only)
     */
    async endRoom(roomName) {
        try {
            const response = await this.apiClient.delete(`/api/livekit/room/${roomName}`);
            return response.data.message;
        } catch (error) {
            console.error('Error ending room:', error);
            throw new Error(`Failed to end room: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Get LiveKit configuration optimized for mobile and iPhone Safari
     */
    async getLiveKitConfig() {
        try {
            const response = await this.apiClient.get('/api/livekit/config');
            return response.data;
        } catch (error) {
            console.error('Error getting LiveKit config:', error);
            // Return fallback config
            return this.getFallbackConfig();
        }
    }

    /**
     * Fallback configuration for LiveKit
     */
    getFallbackConfig() {
        return {
            url: LIVEKIT_URL,
            simulcast: true,
            dynacast: true,
            adaptiveStream: true,
            videoCodecs: {
                preferred: ["h264", "vp9", "av1"],
                backup: ["vp8"]
            },
            audioCodecs: {
                preferred: ["opus"],
                sampleRate: 48000,
                channels: 2
            },
            videoSettings: {
                resolution: {
                    width: 1920,
                    height: 1080
                },
                frameRate: 30,
                bitrate: {
                    min: 500000,
                    max: 5000000,
                    target: 3500000
                }
            },
            audioSettings: {
                echoCancellation: true,
                noiseSuppression: true,
                autoGainControl: true,
                bitrate: 128000
            }
        };
    }

    /**
     * Remove participant from room (admin only)
     */
    async removeParticipant(roomName, participantIdentity) {
        try {
            const response = await this.apiClient.post(
                `/api/livekit/room/${roomName}/participant/${participantIdentity}/remove`
            );
            return response.data.message;
        } catch (error) {
            console.error('Error removing participant:', error);
            throw new Error(`Failed to remove participant: ${error.response?.data?.detail || error.message}`);
        }
    }

    /**
     * Get optimal video constraints for iPhone Safari
     */
    getOptimalVideoConstraints() {
        const isIOS = /iPad|iPhone|iPod/.test(navigator.userAgent) && !window.MSStream;
        const isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
        
        if (isIOS || isSafari) {
            return {
                width: { ideal: 1920, max: 1920 },
                height: { ideal: 1080, max: 1080 },
                frameRate: { ideal: 30, max: 30 },
                facingMode: 'user', // Front camera by default
                aspectRatio: { ideal: 16/9 }
            };
        }

        return {
            width: { ideal: 1920 },
            height: { ideal: 1080 },
            frameRate: { ideal: 30 },
            aspectRatio: { ideal: 16/9 }
        };
    }

    /**
     * Get optimal audio constraints
     */
    getOptimalAudioConstraints() {
        return {
            echoCancellation: true,
            noiseSuppression: true,
            autoGainControl: true,
            sampleRate: 48000,
            channelCount: 2
        };
    }

    /**
     * Create room name with timestamp
     */
    generateRoomName(prefix = 'live-shopping') {
        const timestamp = Date.now();
        return `${prefix}-${timestamp}`;
    }

    /**
     * Check if device supports required features
     */
    async checkDeviceCapabilities() {
        const capabilities = {
            webrtc: !!window.RTCPeerConnection,
            getUserMedia: !!(navigator.mediaDevices && navigator.mediaDevices.getUserMedia),
            camera: false,
            microphone: false,
            h264: false
        };

        try {
            // Check camera and microphone
            const stream = await navigator.mediaDevices.getUserMedia({ video: true, audio: true });
            capabilities.camera = stream.getVideoTracks().length > 0;
            capabilities.microphone = stream.getAudioTracks().length > 0;
            
            // Stop test stream
            stream.getTracks().forEach(track => track.stop());

            // Check H.264 support
            if (window.RTCRtpSender && RTCRtpSender.getCapabilities) {
                const videoCapabilities = RTCRtpSender.getCapabilities('video');
                capabilities.h264 = videoCapabilities.codecs.some(
                    codec => codec.mimeType.toLowerCase().includes('h264')
                );
            }
        } catch (error) {
            console.warn('Error checking device capabilities:', error);
        }

        return capabilities;
    }

    /**
     * Get connection quality metrics
     */
    async getConnectionQuality(room) {
        if (!room) return null;

        try {
            // This would be implemented with LiveKit's connection quality API
            // For now, return a placeholder
            return {
                quality: 'good', // good, fair, poor
                rtt: 0,
                bitrate: 0,
                packetLoss: 0,
                fps: 0
            };
        } catch (error) {
            console.error('Error getting connection quality:', error);
            return null;
        }
    }
}

// Export singleton instance
export const livekitService = new LiveKitService();
export default livekitService;

File: backend/livekit_service.py
"""
LiveKit Cloud Token Service
Implements professional video streaming with LiveKit Cloud
"""

import os
import logging
from datetime import timedelta
from typing import Dict, List, Optional, Any
from livekit import api
import asyncio

logger = logging.getLogger(__name__)

class LiveKitTokenService:
    """Service for managing LiveKit Cloud tokens and rooms"""
    
    def __init__(self):
        """Initialize LiveKit service with credentials from environment"""
        self.api_key = os.getenv("LIVEKIT_API_KEY")
        self.api_secret = os.getenv("LIVEKIT_API_SECRET")
        self.livekit_url = os.getenv("LIVEKIT_URL", "wss://live-stream-q7s7lvvw.livekit.cloud")
        
        if not self.api_key or not self.api_secret:
            logger.warning("LiveKit API credentials not configured - LiveKit features will be disabled")
            self.livekit_api = None
            return
        
        try:
            # Initialize LiveKit API client
            self.livekit_api = api.LiveKitAPI(
                url=self.livekit_url,
                api_key=self.api_key,
                api_secret=self.api_secret
            )
            
            logger.info(f"LiveKit service initialized with URL: {self.livekit_url}")
        except Exception as e:
            logger.error(f"Failed to initialize LiveKit API: {str(e)}")
            self.livekit_api = None
    
    async def create_publisher_token(
        self,
        room_name: str,
        participant_identity: str,
        participant_name: str = None,
        metadata: Dict[str, Any] = None
    ) -> str:
        """
        Generate LiveKit access token for publisher (admin streamer)
        
        Args:
            room_name: Name of the room
            participant_identity: Unique identity for participant
            participant_name: Display name for participant
            metadata: Additional metadata
            
        Returns:
            JWT token string for LiveKit access
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            # Create access token with publisher permissions
            token = api.AccessToken(self.api_key, self.api_secret)
            token.with_identity(participant_identity)
            
            if participant_name:
                token.with_name(participant_name)
            
            if metadata:
                token.with_metadata(str(metadata))
            
            # Configure video grants for publisher (full permissions)
            video_grants = api.VideoGrants(
                room_join=True,
                room=room_name,
                can_publish=True,        # Can publish video/audio
                can_subscribe=True,      # Can subscribe to other streams
                can_publish_data=True,   # Can send chat messages
                can_update_own_metadata=True, # Can update own metadata
                # High quality settings for publisher
                room_admin=True,
                room_create=True
            )
            
            token.with_grants(video_grants)
            
            # Token valid for 2 hours (streaming session)
            token.with_ttl(timedelta(hours=2))
            
            jwt_token = token.to_jwt()
            logger.info(f"Publisher token created for {participant_identity} in room {room_name}")
            
            return jwt_token
            
        except Exception as e:
            logger.error(f"Failed to create publisher token: {str(e)}")
            raise Exception(f"Token generation failed: {str(e)}")
    
    async def create_viewer_token(
        self,
        room_name: str,
        participant_identity: str,
        participant_name: str = None,
        metadata: Dict[str, Any] = None
    ) -> str:
        """
        Generate LiveKit access token for viewer (view-only customer)
        
        Args:
            room_name: Name of the room
            participant_identity: Unique identity for participant
            participant_name: Display name for participant
            metadata: Additional metadata
            
        Returns:
            JWT token string for LiveKit access
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            # Create access token with viewer permissions only
            token = api.AccessToken(self.api_key, self.api_secret)
            token.with_identity(participant_identity)
            
            if participant_name:
                token.with_name(participant_name)
            
            if metadata:
                token.with_metadata(str(metadata))
            
            # Configure video grants for viewer (view-only permissions)
            video_grants = api.VideoGrants(
                room_join=True,
                room=room_name,
                can_publish=False,       # Cannot publish video/audio
                can_subscribe=True,      # Can subscribe to streams
                can_publish_data=True,   # Can send chat messages
                can_update_own_metadata=False, # Cannot update metadata
                room_admin=False,
                room_create=False
            )
            
            token.with_grants(video_grants)
            
            # Token valid for 2 hours
            token.with_ttl(timedelta(hours=2))
            
            jwt_token = token.to_jwt()
            logger.info(f"Viewer token created for {participant_identity} in room {room_name}")
            
            return jwt_token
            
        except Exception as e:
            logger.error(f"Failed to create viewer token: {str(e)}")
            raise Exception(f"Token generation failed: {str(e)}")
    
    async def create_room(
        self,
        room_name: str,
        max_participants: int = 50,
        empty_timeout: int = 300,
        metadata: Dict[str, Any] = None
    ) -> Dict[str, Any]:
        """
        Create a new LiveKit room for streaming
        
        Args:
            room_name: Name of the room to create
            max_participants: Maximum number of participants
            empty_timeout: Timeout when room is empty (seconds)
            metadata: Room metadata
            
        Returns:
            Room information dictionary
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            # Create room request
            room_request = api.CreateRoomRequest(
                name=room_name,
                empty_timeout=empty_timeout,
                max_participants=max_participants,
                metadata=str(metadata) if metadata else ""
            )
            
            # Create the room
            room_info = await self.livekit_api.room.create_room(room_request)
            
            logger.info(f"Room created: {room_name} with max participants: {max_participants}")
            
            return {
                "name": room_info.name,
                "sid": room_info.sid,
                "creation_time": room_info.creation_time,
                "max_participants": room_info.max_participants,
                "num_participants": room_info.num_participants,
                "empty_timeout": room_info.empty_timeout,
                "metadata": room_info.metadata
            }
            
        except Exception as e:
            logger.error(f"Failed to create room {room_name}: {str(e)}")
            raise Exception(f"Room creation failed: {str(e)}")
    
    async def get_room_info(self, room_name: str) -> Optional[Dict[str, Any]]:
        """
        Get information about a specific room
        
        Args:
            room_name: Name of the room
            
        Returns:
            Room information or None if room doesn't exist
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            # List rooms and find the specific one
            request = api.ListRoomsRequest()
            response = await self.livekit_api.room.list_rooms(request)
            
            for room in response.rooms:
                if room.name == room_name:
                    return {
                        "name": room.name,
                        "sid": room.sid,
                        "num_participants": room.num_participants,
                        "max_participants": room.max_participants,
                        "creation_time": room.creation_time,
                        "metadata": room.metadata,
                        "empty_timeout": room.empty_timeout
                    }
            
            return None
            
        except Exception as e:
            logger.error(f"Failed to get room info for {room_name}: {str(e)}")
            raise Exception(f"Failed to get room info: {str(e)}")
    
    async def list_active_rooms(self) -> List[Dict[str, Any]]:
        """
        List all active LiveKit rooms
        
        Returns:
            List of room information dictionaries
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            request = api.ListRoomsRequest()
            response = await self.livekit_api.room.list_rooms(request)
            
            rooms = []
            for room in response.rooms:
                rooms.append({
                    "name": room.name,
                    "sid": room.sid,
                    "num_participants": room.num_participants,
                    "max_participants": room.max_participants,
                    "creation_time": room.creation_time,
                    "metadata": room.metadata
                })
            
            logger.info(f"Listed {len(rooms)} active rooms")
            return rooms
            
        except Exception as e:
            logger.error(f"Failed to list rooms: {str(e)}")
            raise Exception(f"Failed to list rooms: {str(e)}")
    
    async def end_room(self, room_name: str) -> bool:
        """
        End/delete a LiveKit room
        
        Args:
            room_name: Name of the room to end
            
        Returns:
            True if successful, False otherwise
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            request = api.DeleteRoomRequest(room=room_name)
            await self.livekit_api.room.delete_room(request)
            
            logger.info(f"Room ended: {room_name}")
            return True
            
        except Exception as e:
            logger.error(f"Failed to end room {room_name}: {str(e)}")
            return False
    
    async def get_participants(self, room_name: str) -> List[Dict[str, Any]]:
        """
        Get list of participants in a room
        
        Args:
            room_name: Name of the room
            
        Returns:
            List of participant information
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            request = api.ListParticipantsRequest(room=room_name)
            response = await self.livekit_api.room.list_participants(request)
            
            participants = []
            for participant in response.participants:
                participants.append({
                    "identity": participant.identity,
                    "name": participant.name,
                    "sid": participant.sid,
                    "state": participant.state.name,
                    "joined_at": participant.joined_at,
                    "metadata": participant.metadata,
                    "permission": {
                        "can_subscribe": participant.permission.can_subscribe,
                        "can_publish": participant.permission.can_publish,
                        "can_publish_data": participant.permission.can_publish_data
                    }
                })
            
            return participants
            
        except Exception as e:
            logger.error(f"Failed to get participants for room {room_name}: {str(e)}")
            return []
    
    async def remove_participant(
        self,
        room_name: str,
        participant_identity: str
    ) -> bool:
        """
        Remove a participant from a room
        
        Args:
            room_name: Name of the room
            participant_identity: Identity of participant to remove
            
        Returns:
            True if successful, False otherwise
        """
        if not self.livekit_api:
            raise Exception("LiveKit service not properly initialized - credentials missing")
            
        try:
            request = api.RemoveParticipantRequest(
                room=room_name,
                identity=participant_identity
            )
            
            await self.livekit_api.room.remove_participant(request)
            logger.info(f"Participant {participant_identity} removed from room {room_name}")
            
            return True
            
        except Exception as e:
            logger.error(f"Failed to remove participant {participant_identity}: {str(e)}")
            return False
    
    async def get_livekit_config(self) -> Dict[str, Any]:
        """
        Get LiveKit configuration for client-side connection
        
        Returns:
            Configuration dictionary for LiveKit client
        """
        return {
            "url": self.livekit_url,
            "simulcast": True,  # Enable simulcast for quality adaptation
            "dynacast": True,   # Enable dynacast for efficient streaming
            "adaptiveStream": True,  # Enable adaptive streaming
            "videoCodecs": {
                "preferred": ["h264", "vp9", "av1"],  # Codec priority
                "backup": ["vp8"]
            },
            "audioCodecs": {
                "preferred": ["opus"],  # High quality Opus audio
                "sampleRate": 48000,    # 48 kHz sample rate
                "channels": 2           # Stereo
            },
            "videoSettings": {
                "resolution": {
                    "width": 1920,      # 1080p resolution
                    "height": 1080
                },
                "frameRate": 30,        # 30fps
                "bitrate": {
                    "min": 500000,      # 0.5 Mbps minimum
                    "max": 5000000,     # 5 Mbps maximum
                    "target": 3500000   # 3.5 Mbps target
                }
            },
            "audioSettings": {
                "echoCancellation": True,
                "noiseSuppression": True,
                "autoGainControl": True,
                "bitrate": 128000       # 128 kbps audio
            }
        }
    
    async def close(self):
        """Close the LiveKit API connection"""
        if not self.livekit_api:
            return
            
        try:
            await self.livekit_api.aclose()
            logger.info("LiveKit service connection closed")
        except Exception as e:
            logger.error(f"Error closing LiveKit service: {str(e)}")

# Global instance
livekit_service = LiveKitTokenService()